### PR TITLE
Merge base values with overlay values  

### PR DESCRIPTION
**Pull request by jala-dx**
Merge base values with overlay values if the keys are unique. if the keys conflict, overlay takes precendence.

if nothing is specified in base and overlay, values.yaml takes effect

### CODE REVIEW

File: kotskinds/apis/kots/v1beta1/helmchart_types.go [hunk 1]
```diff
@@ -162,18 +163,107 @@ type ChartIdentifier struct {
...
+	for k, v := range overlayValues {
+		if _, exists := baseValues[k]; !exists {
+			result[k] = v
+		}
+	}
+	return result
+
+}
+
+func RecurseMerge(baseValues map[string]*MappedChartValue, overlayValues map[string]*MappedChartValue) map[string]*MappedChartValue {
```
Comment (divolgin) [inline kotskinds/apis/kots/v1beta1/helmchart_types.go:221]: ~RecurseMerge~ `mergeValueChildren`

File: kotskinds/apis/kots/v1beta1/helmchart_types.go [hunk 2]
```diff
@@ -174,6 +175,85 @@ func (h *HelmChartSpec) GetHelmValues(values map[string]MappedChartValue) (map[s
...
+	for k, v := range overlayValues {
+		if _, exists := baseValues[k]; !exists {
+			result[k] = v
+		}
+	}
+	return result
+
+}
+
+func PrintResultMap(baseValues map[string]MappedChartValue) {
```
Comment (jala-dx) [inline kotskinds/apis/kots/v1beta1/helmchart_types.go:238]: Remove this func before merging

File: kotskinds/apis/kots/v1beta1/helmchart_types.go [hunk 3]
```diff
@@ -174,6 +175,85 @@ func (h *HelmChartSpec) GetHelmValues(values map[string]MappedChartValue) (map[s
...
+
+	for k, v := range baseValues {
+		if v.valueType == "children" {
+			RecursePrint(v.children)
+		}
+		fmt.Println("PrintResultMap+++++++++++++key, value\n", k, v)
+	}
+}
+
+func RecursePrint(baseValues map[string]*MappedChartValue) {
```
Comment (jala-dx) [inline kotskinds/apis/kots/v1beta1/helmchart_types.go:248]: Remove this func before merg

File: kotskinds/apis/kots/v1beta1/helmchart_types_test.go [hunk 4]
```diff
@@ -339,3 +342,183 @@ func Test_MappedChartValueGetValue(t *testing.T) {
...
+					valueType: "children",
+					children: map[string]*MappedChartValue{
+						"postgres": &MappedChartValue{
+							valueType: "children",
+							children: map[string]*MappedChartValue{
+								"enabled": &MappedChartValue{
+									boolValue: true,
+									valueType: "bool",
+								},
+								"replacementtest": &MappedChartValue{
```
Comment (emosbaugh) [inline kotskinds/apis/kots/v1beta1/helmchart_types_test.go:474]: Isn't the reason for this change when there is a deep merge and the overlay does not contain the key from the base? Or am I missing something? It seems you're missing a test for that case.

Comment (jala-dx) [reply kotskinds/apis/kots/v1beta1/helmchart_types_test.go:474]: Nice thanks, added the test case

File: pkg/base/replicated.go [hunk 5]
```diff
@@ -137,12 +137,14 @@ func renderReplicated(u *upstreamtypes.Upstream, renderOptions *RenderOptions) (
...
-				mergedValues[k] = v
+			if optionalValues.RecursiveMerge {
+				mergedValues = kotsv1beta1.MergeHelmChartValues(kotsHelmChart.Spec.Values, optionalValues.Values)
+			} else {
+				for k, v := range optionalValues.Values {
+					mergedValues[k] = v
+				}
 			}
 		}
-
```
Comment (divolgin) [inline pkg/base/replicated.go:145]: would be cool to keep this line

**Review by divolgin — APPROVED** [review]

File: kotskinds/apis/kots/v1beta1/helmchart_types_test.go [hunk 6]
```diff
@@ -339,3 +342,235 @@ func Test_MappedChartValueGetValue(t *testing.T) {
...
+				},
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			actual := MergeHelmChartValues(test.baseValues, test.overlayValues)
+			deep.CompareUnexportedFields = true
+			diff := deep.Equal(&actual, &test.expect)
```
Comment (salahalsaleh) [inline kotskinds/apis/kots/v1beta1/helmchart_types_test.go:569]: why not use `reflect.DeepEqual`? no need for another library

File: pkg/base/replicated.go [hunk 7]
```diff
@@ -137,9 +137,12 @@ func renderReplicated(u *upstreamtypes.Upstream, renderOptions *RenderOptions) (
 			if !parsedBool {
 				continue
 			}
-
-			for k, v := range optionalValues.Values {
-				mergedValues[k] = v
+			if optionalValues.RecursiveMerge {
+				mergedValues = kotsv1beta1.MergeHelmChartValues(kotsHelmChart.Spec.Values, optionalValues.Values)
```
Comment (salahalsaleh) [inline pkg/base/replicated.go:141]: `optionalValues` are not always coming from `overlays`, right? i guess i'm confused by the param name `overlayValues` here https://github.com/replicatedhq/kots/pull/1751/files#diff-6f2a6d88643d24717b68bd9924433c1f9f1f4b44567cce0b1b0db8b47486c270R177

File: kotskinds/apis/kots/v1beta1/helmchart_types_test.go (hunk 6 above)
Comment (dexhorthy) [reply kotskinds/apis/kots/v1beta1/helmchart_types_test.go:569]: I had seen deep.Equal used in many other projects -- I think the biggest advantage is that it prints a pretty diff

Comment (dexhorthy) [reply kotskinds/apis/kots/v1beta1/helmchart_types_test.go:569]: @laverya was the last one I used it w/ -- are we still into this library?

File: kotskinds/apis/kots/v1beta1/helmchart_types.go [hunk 8]
```diff
@@ -174,6 +174,57 @@ func (h *HelmChartSpec) GetHelmValues(values map[string]MappedChartValue) (map[s
...
+	overlayValues map[string]MappedChartValue) map[string]MappedChartValue {
+
+	result := map[string]MappedChartValue{}
+	for k, v := range baseValues {
+		if _, exists := overlayValues[k]; !exists {
+			result[k] = baseValues[k]
+			continue
+		}
+		if v.valueType != "children" {
+			result[k] = overlayValues[k]
```
Comment (emosbaugh) [inline kotskinds/apis/kots/v1beta1/helmchart_types.go:187]: can you have an array of maps?

Comment (emosbaugh) [reply kotskinds/apis/kots/v1beta1/helmchart_types.go:187]: never mind. bad question

File: kotskinds/apis/kots/v1beta1/helmchart_types.go [hunk 9]
```diff
@@ -174,6 +174,57 @@ func (h *HelmChartSpec) GetHelmValues(values map[string]MappedChartValue) (map[s
...
+	}
+	for k, v := range overlayValues {
+		if _, exists := baseValues[k]; !exists {
+			result[k] = v
+		}
+	}
+	return result
+}
+
+func mergeValueChildren(baseValues map[string]*MappedChartValue, overlayValues map[string]*MappedChartValue) map[string]*MappedChartValue {
```
Comment (emosbaugh) [inline kotskinds/apis/kots/v1beta1/helmchart_types.go:203]: is this the same (or almost the same) as the function above? if so why duplicate?

**Review by emosbaugh — APPROVED** [review]
generally approve. just one comment about code cleanup

**Review by salahalsaleh — APPROVED** [review]

File: kotskinds/apis/kots/v1beta1/helmchart_types.go (hunk 9 above)
Comment (jala-dx) [reply kotskinds/apis/kots/v1beta1/helmchart_types.go:203]: yes i can explain, the top level struct passed is map[string]MappedChartValue, but the children is map[string]*MappedChartValue.  unless have two functions, it was not possible to recurse.

**Pull request merged**